### ISSUE TITLE
Crypto - KeyObject is undefined for Node V14 - "jsonwebtoken": "9.0.0"

### ISSUE THREAD

**Issue by josephnvn** [auth0/node-jsonwebtoken#885]
**Please do not report security vulnerabilities here**. The [Responsible Disclosure Program](https://auth0.com/whitehat) details the procedure for disclosing security issues.

**Thank you in advance for helping us to improve this library!** Please read through the template below and answer all relevant questions. Your additional work here is greatly appreciated and will help us respond as quickly as possible. For general support or usage questions, use the [Auth0 Community](https://community.auth0.com/) or [Auth0 Support](https://support.auth0.com/). Finally, to avoid duplicates, please search existing Issues before submitting one here.

By submitting an Issue to this repository, you agree to the terms within the [Auth0 Code of Conduct](https://github.com/auth0/open-source-template/blob/master/CODE-OF-CONDUCT.md).

### Description
    KeyObject from Crypto is undefined. I am using Node 14.10.1. 

### Reproduction

> NODE v14.10.1
   jwt.sign({email: "bar@foo"}, "PRIVATE_KEY", { algorithm: 'HS512' })
   
> Throws an error
   > webpack-internal:///…ebtoken/sign.js:173 Uncaught (in promise) TypeError: Right-hand side of 'instanceof' is not an object
    at module.exports [as sign] (sign.js:108:1)
    `if (secretOrPrivateKey != null && !(secretOrPrivateKey instanceof KeyObject)) {`
   
<img width="1039" alt="image" src="https://user-images.githubusercontent.com/1001333/217306933-f06bad3a-263d-48aa-8759-29c01115108c.png">

### Environment

> Please provide the following:

- **Version of this library used:** 9.0.0
- **Version of the platform or framework used, if applicable:** any platform
- **Other relevant versions (language, server software, OS, browser):** 
- **Other modules/plugins/libraries that might be involved:**

**Comment by declaresub** [comment]
I tripped over the same issue.  I believe the problem is that KeyObject belongs to node Crypto, and so does not exist in the browser environment.

**Comment by panva** [comment]
Indeed, `KeyObject` is [present in Node.js v14.10.1](https://nodejs.org/docs/v14.10.1/api/crypto.html#crypto_class_keyobject). This library's use outside of Node.js runtime was never intended. 

Versions before 9.0.0 could use webpack's polyfill of the Node.js built-in modules. This most likely a) relied on slow js cryptography that isn't [maintained anymore](https://www.npmjs.com/package/crypto-browserify) and lacks feature parity with Node's crypto and b) increased your js bundle size considerably. FWIW webpack 5 also no longer defaults to using these questionable crypto polyfills anymore.

It is better to rely on JWT modules made to be consumed in browser or generic javascript environments. You can discover such modules on [jwt.io](https://jwt.io/libraries?language=JavaScript) under "JavaScript" libraries.

**Comment by dhwang** [comment]
[Node 15 implementation adds WebCrypto and CryptoKey](https://github.com/panva/node/commit/dae283d96fd31ad0f30840a7e55ac97294f505ac#diff-dc0c77ec6254d6ea006052823a4d90dda37543cc4c4bf451a6fd8d079b9fa9f3)
and has implementation for both `isCryptoKey` and `isKeyObject`

What blocks jsonwebtoken from creating similar  `isCryptoKey` and `isKeyObject` and test both?

This library has been is much smaller package for its purpose and we like to keep using it instead of the alternatives.

**Comment by fandrusi** [comment]
Running into this same issue :(

**Comment by rachellerathbone** [comment]
This is an issue in 9.0.1 as well. I'm migrating from 8.5.1, the last version before the major version bump, because 8.5.1 uses semver@5.6.0 which has a known ReDoS vulnerability.

According to your migration [docs](https://github.com/auth0/node-jsonwebtoken/wiki/Migration-Notes:-v8-to-v9), to migration from 8 > 9 I only need to:

- ensure I'm running on a version of Node 12 or higher (using version 16.19.1)
- pass `{ algorithms: ['none'] }` to the verify function for unsigned tokens. 

However, there seems to be a number of failure points in 9.0.x:

1. Line #110 - `secretOrPrivateKey != null && !(secretOrPrivateKey instanceof KeyObject` - in verify.js
The type for secretOrPrivateKey is `Secret` which is any of the following:

```
export type Secret =
    | string
    | Buffer
    | { key: string | Buffer; passphrase: string };
```

You can't call instance of on anything except an object otherwise you end up with `TypeError: Right-hand side of 'instanceof' is not an object`. This could be easily resolved by wrapping the conditional block where this call is made in an additional check e.g. `typeof secretOrPublicKey === "object`

Unfortunately, there are more failure points beyond this.

2. Lines #148-150 in verify js
```
if (options.algorithms.indexOf(decodedToken.header.alg) === -1) {
      return done(new JsonWebTokenError('invalid algorithm'));
    }
```

Despite your migration docs stating you can simply pass 'none' to the algorithms array, the value of `decodedToken.header.alg` is HS256 (HMAC using SHA-256 hash algorithm). So obviously, when you pass `{ algorithms: ['none'] }` this fails this check and spits out the error `invalid algorithm`. Even if this conditional is updated to check if 'none' is being passed to algorithms, it then fails on line #152 `if (header.alg.startsWith('HS') && secretOrPublicKey.type !== 'secret')` because header.alg is HS256. Then, if you resolve this error, the next one you hit is `invalid signature` because verify, set on line #169 `valid = jws.verify(jwtString, decodedToken.header.alg, secretOrPublicKey);` returns false (again because decodedToken.header.alg is defaulting to the default value and not honouring 'none').

However, if you add another conditional check to bypass the validation conditional if 'none' is passed, it succeeds without any errors.

Can someone look at updating this ASAP? Being stuck with a version that has a pretty serious VULN definitely isn't ideal.

**Comment by rachellerathbone** [comment]
This is an issue in 9.0.1 as well. I'm migrating from 8.5.1, the last version before the major version bump, because 8.5.1 uses semver@5.6.0 which has a known ReDoS vulnerability.

According to your migration docs, to migration from 8 > 9 I only need to:
* ensure I'm running on a version of Node 12 or higher (using version 16.19.1)
* pass { algorithms: ['none'] } to the verify function for unsigned tokens.

However, there seems to be a number of failure points in 9.0.x:

1. Line #110 - `secretOrPrivateKey != null && !(secretOrPrivateKey instanceof KeyObject` - in verify.js

The type for secretOrPrivateKey is Secret which is any of the following:
```
export type Secret =
    | string
    | Buffer
    | { key: string | Buffer; passphrase: string };
```    
    
You can't call instance of on anything except an object otherwise you end up with TypeError: Right-hand side of 'instanceof' is not an object. This could be easily resolved by wrapping the conditional block where this call is made in an additional check e.g. typeof secretOrPublicKey === "object

Unfortunately, there are more failure points beyond this.

2. Lines  #148-150 in verify js
```
if (options.algorithms.indexOf(decodedToken.header.alg) === -1) {
      return done(new JsonWebTokenError('invalid algorithm'));
    }
 ```
 
Despite your migration docs stating you can simply pass 'none' to the algorithms array, the value of decodedToken.header.alg is HS256 (HMAC using SHA-256 hash algorithm). So obviously, when you pass { algorithms: ['none'] } this fails this check and spits out the error invalid algorithm. Even if this conditional is updated to check if 'none' is being passed to algorithms, it then fails on line #152 `if (header.alg.startsWith('HS') && secretOrPublicKey.type !== 'secret')` because header.alg is HS256. Then, if you resolve this error, the next one you hit is invalid signature because verify, set on line #169 valid = jws.verify(jwtString, decodedToken.header.alg, secretOrPublicKey); returns false (again because decodedToken.header.alg is defaulting to the default value and not honouring 'none').
However, if you add another conditional check to bypass the validation conditional if 'none' is passed, it succeeds without any errors.

Can someone look at updating this ASAP? Being stuck with a version that has a pretty serious VULN definitely isn't ideal.

**Comment by WeiRU087** [comment]
Any solution around this?

**Comment by siraj-iconcile** [comment]
Is this repo even being maintained? 
I see multiple issues open for that one thing and no resolution left. 8.x version has vulnerabilities.  And 9.x giving this error.
